src/providers/wfs/qgswfsrequest.cpp

```diff
@@ -201,20 +201,20 @@ bool QgsWfsRequest::sendGET( const QUrl &url, bool synchronous, bool forceRefres
       // he needs the authentication to run.
       // The processEvents() call gives the auth manager the chance to show a dialog and
       // once done with that, we can wake the downloaderThread again and continue the download.
-      if ( !downloaderThread->isFinished() )
+      if ( !threadFinished )
```

Comment:
the new state should probably be read under the mutex
(I doubt that this remark and the above ones are actually going to fix real-world problems as some of the mechanisms above probably already act as full-memory barrier, but this is just for the sake of mind relief)